File: 프로그래머스/3/60061. 기둥과 보 설치/기둥과 보 설치.py
def solution(n, build_frame):
    answer = []
    
    def isOk():
        for x,y,a in answer:
            if a == 0: #기둥
                if y != 0 and [x,y,1] not in answer and [x-1,y,1] not in answer and [x,y-1,0] not in answer:
                    return False
            else:
                if [x,y-1,0] not in answer and [x+1,y-1,0] not in answer and not([x-1,y,1] in answer and [x+1,y,1] in answer):
                    return False
                
        return True
    
    for x,y,a,b in build_frame:
        if b == 1: #설치
            answer.append([x,y,a])
            if not isOk():
                answer.remove([x,y,a])
        else: #  삭제
            answer.remove([x,y,a])
            if not isOk():
                answer.append([x,y,a])
    
    answer.sort(key =lambda x:(x[0],x[1],x[2]))
    
    return answer

File: 프로그래머스/3/60061. 기둥과 보 설치/기둥과 보 설치.js
function solution(n, build_frame) {
    const answer = [];
    
    function isOk() {
        for (const [x, y, a] of answer) {
            if (a === 0) { // Column
                if (
                    y !== 0 &&
                    !answer.some(([px, py, pa]) => px === x && py === y - 1 && pa === 0) &&
                    !answer.some(([px, py, pa]) => px === x - 1 && py === y && pa === 1) &&
                    !answer.some(([px, py, pa]) => px === x && py === y && pa === 1)
                ) {
                    return false;
                }
            } else { // Beam
                if (
                    !answer.some(([px, py, pa]) => px === x && py === y - 1 && pa === 0) &&
                    !answer.some(([px, py, pa]) => px === x + 1 && py === y - 1 && pa === 0) &&
                    !(
                        answer.some(([px, py, pa]) => px === x - 1 && py === y && pa === 1) &&
                        answer.some(([px, py, pa]) => px === x + 1 && py === y && pa === 1)
                    )
                ) {
                    return false;
                }
            }
        }
        return true;
    }
    
    for (const [x, y, a, b] of build_frame) {
        if (b === 1) { // Install
            answer.push([x, y, a]);
            if (!isOk()) {
                answer.pop();
            }
        } else { // Delete
            const index = answer.findIndex(([px, py, pa]) => px === x && py === y && pa === a);
            if (index > -1) {
                answer.splice(index, 1);
                if (!isOk()) {
                    answer.push([x, y, a]);
                }
            }
        }
    }
    
    answer.sort((a, b) => a[0] - b[0] || a[1] - b[1] || a[2] - b[2]);
    
    return answer;
    return answer;
}